src/Workspaces/Core/Portable/Shared/Extensions/ContextQuery/AbstractSyntaxContext.cs

```diff
@@ -80,21 +82,27 @@ private ISet<INamedTypeSymbol> ComputeOuterTypes(CancellationToken cancellationT
                 var containingType = enclosingSymbol.GetContainingTypeOrThis();
                 if (containingType != null)
                 {
-                    return containingType.GetContainingTypes().ToSet();
+                    return new Tuple<ISet<ITypeSymbol>, ITypeSymbol>(containingType.GetContainingTypesAndThis().ToSet(), containingType);
                 }
             }
 
-            return SpecializedCollections.EmptySet<INamedTypeSymbol>();
+            return new Tuple<ISet<ITypeSymbol>, ITypeSymbol>(SpecializedCollections.EmptySet<ITypeSymbol>(), null);
         }
 
-        public ISet<INamedTypeSymbol> GetOuterTypes(CancellationToken cancellationToken)
+        public IEnumerable<ITypeSymbol> GetOuterTypes(bool includeContainingType, CancellationToken cancellationToken)
         {
-            if (_outerTypes == null)
+            if (_outerTypesIncludingContainingTypeAndContainingType == null)
             {
-                Interlocked.CompareExchange(ref _outerTypes, ComputeOuterTypes(cancellationToken), null);
+                Interlocked.CompareExchange(ref _outerTypesIncludingContainingTypeAndContainingType, ComputeOuterTypes(cancellationToken), null);
+            }
+            var outerTypesIncludingContainingType = _outerTypesIncludingContainingTypeAndContainingType.Item1;
+            var containingType = _outerTypesIncludingContainingTypeAndContainingType.Item2;
+            if (includeContainingType)
+            {
+                return outerTypesIncludingContainingType;
             }
 
-            return _outerTypes;
+            return outerTypesIncludingContainingType.Where(x => x != containingType);
```

Comment:
This seems really bizzare to me. My suggestion would be to not change the semantics of `GetOuterTypes(...)`. Instead, call `GetContainingTypeOrThis(...)` in the recommendation service and then call `GetOuterTypes(...)`. Also, in `AbstractRecommendationService` there's a method called `GetEnclosingTypeBases(...)` that might serve you better. You could try making that protected and calling it instead.